### ISSUE TITLE
--folder option doesn't allow for whitespace fixes.

### ISSUE THREAD

**Issue by Tylertron1998** [dotnet/format#878]
I've just stumbled across this tool and am wondering about using it for some basic chat bot formatting in discord - however, it doesn't allow for `--fix-whitespace` when using the `--folder` option. Is this on purpose? Or is there any way around this?

**Comment by Tylertron1998** [comment]
scratch that - looks like that option has been removed in later builds anyway. However I am still running into an issue when trimming whitespace:

given the input of:
```csharp

using System;

for (var i = 0; i < 10010100; i++)
{



    Console.WriteLine("HELLO WORLD");



}
```

I would expect the output of:

```csharp
using System;

for (var i = 0; i < 10010100; i++)
{
    Console.WriteLine("HELLO WORLD");
}
```
however it doesn't seem to quite work.

**Comment by JoeRobich** [comment]
@Tylertron1998 Glad things are working a little better for you. The Roslyn formatter does not have a rule for removing blank lines. There is a StyleCop Analyzer package which you can add to your project which will provide an analyzer and codefix for consecutive blank lines (See this comment https://github.com/dotnet/format/issues/67#issuecomment-478958324). The nightly 5.x builds are able to run 3rd party analyzers and fixers (See https://github.com/dotnet/format/tree/master/docs#3rd-party-analysis). See how to install the 5.x build here - https://github.com/dotnet/format#development-builds-v5x

Closing this issue as a duplicate of  https://github.com/dotnet/format/issues/67

**Comment by Tylertron1998** [comment]
> Running 3rd party analysis requires the use of a MSBuild solution or project file as the workspace. 3rd party analyzers are discovered from the <PackageReferences> specified in the workspace project files.

Hm. This is rather unfortunate as this means I'd *need* a workspace, right? That's a bit of a problem, as my exact usecase for this is for formatting small snippets, that are sent in the C# discord - so the snippets aren't always 'valid' C# in the sense that someone might just send a method, or the body of a method - for example. What I am currently doing is something like this:

- Take in the code i.e.:

```csharp
for(var i = 0; i < 10; i++)

{
Console.WriteLine( "Hello" + i ); }
```
- writing the code to a file (test.cs)
- having the formatter 'correct' the file
- reading its contents, and sending it back.

Now, from what I gather, the formatter won't actually work for this in context of a project/workspace, as the code would have to be a 100% valid C# file - is this correct?

**Comment by JoeRobich** [comment]
@Tylertron1998  I put together a proof of concept for your example. The C# compiler is very good about providing semantic information for code that may not be fully compilable. In addition, the new support for Top Level Statements probably helps here as well (see https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/exploration/top-level-statements).

While looking into this exact set of analyzers and fixers, I've fixed a dotnet format bug that was keeping StyleCop fixes from being applied (https://github.com/dotnet/format/pull/884).

toplevel.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="StyleCop.Analyzers" Version="1.1.118">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
    </PackageReference>
  </ItemGroup>
</Project>
```

.editorconfig
```ini
root = true

[*.cs]
indent_style = space
indent_size = 4
insert_final_newline = false

# Disable all StyleCop analyzers then we can enable them one at a time.
dotnet_analyzer_diagnostic.category-StyleCop.CSharp.DocumentationRules.severity = none
dotnet_analyzer_diagnostic.category-StyleCop.CSharp.LayoutRules.severity = none
dotnet_analyzer_diagnostic.category-StyleCop.CSharp.MaintainabilityRules.severity = none
dotnet_analyzer_diagnostic.category-StyleCop.CSharp.NamingRules.severity = none
dotnet_analyzer_diagnostic.category-StyleCop.CSharp.OrderingRules.severity = none
dotnet_analyzer_diagnostic.category-StyleCop.CSharp.ReadabilityRules.severity = none
dotnet_analyzer_diagnostic.category-StyleCop.CSharp.SpacingRules.severity = none
dotnet_analyzer_diagnostic.category-StyleCop.CSharp.SpecialRules.severity = none

# Two or more consecutive blank lines: Remove down to one blank line. SA1507
dotnet_diagnostic.SA1507.severity = error

# Blank line immediately before or after a { line: remove it. SA1505, SA1509
dotnet_diagnostic.SA1505.severity = error
dotnet_diagnostic.SA1509.severity = error

# Blank line immediately before a } line: remove it. SA1508
dotnet_diagnostic.SA1508.severity = error
```

Program.cs
```csharp
for(var i = 0; i < 10; i++)

{
Console.WriteLine( "Hello" + i ); }
```

Once the bug fix mentioned above is merged and a new nightly build is available. Running dotnet-format should produce the following output for this project.

```console
~/Source/toplevel> dotnet format --fix-whitespace --fix-analyzers -v diag
  The dotnet runtime version is '5.0.0'.
  The dotnet CLI version is '5.0.100'.
  Using MSBuild.exe located in '/usr/local/share/dotnet/sdk/5.0.100/'.
  Formatting code files in workspace '/Users/joeyrobichaud/Source/toplevel/toplevel.csproj'.
  Loading workspace.
  Complete in 3521ms.
  Determining formattable files.
  Complete in 359ms.
  Running formatters.
  Program.cs(1,1): Fix whitespace formatting.
  Program.cs(5,2): Fix final newline.
  Running Analyzer Reference analysis.
  Determining diagnostics...
  Running 4 analyzers on toplevel.
  Program.cs(4,5): The name 'Console' does not exist in the current context (CS0103)
  Program.cs(3,1): Opening braces should not be preceded by blank line. (SA1509)
  Complete in 1758ms.
  Fixing diagnostics...
  Unable to fix CS0103. No associated code fix found.
  Running 1 analyzers on toplevel.
  Complete in 374ms.
  Analysis complete in 2133ms.
  Complete in 2943ms.
  Formatted code file '/Users/joeyrobichaud/Source/toplevel/Program.cs'.
  Formatted 1 of 3 files.
  Format complete in 6877ms.
```

The resulting Program.cs file will be formatted like so:
```csharp
for (var i = 0; i < 10; i++)
{
    Console.WriteLine("Hello" + i);
}
```

**Comment by JoeRobich** [comment]
Version 5.0.160201 is available and has the StyleCop FixAll fix.

**Comment by Tylertron1998** [comment]
@JoeRobich this is awesome! Thank you so much - this looks perfect.
